File: packages/core/src/providers/registry.ts
import { UnsupportedProviderError } from "../errors.js";
import { AnthropicAdapter } from "./anthropic.js";
import { AzureAdapter } from "./azure.js";
import { GeminiAdapter } from "./gemini.js";
import { OpenAIAdapter } from "./openai.js";
import { PROVIDER_PRESETS } from "./presets.js";
import type { ProviderAdapter } from "./types.js";

const openai = new OpenAIAdapter();
const azure = new AzureAdapter();
const anthropic = new AnthropicAdapter();
const gemini = new GeminiAdapter();

/**
 * Built-in adapters. "openai-compatible" is the same object as "openai" —
 * only the base URL differs.
 */
const builtinAdapters: Record<string, ProviderAdapter> = {
  openai,
  azure,
  anthropic,
  gemini,
};

/** Third-party adapters registered via registerAdapter (factory cache). */
const customAdapters = new Map<string, ProviderAdapter>();

/**
 * Register a third-party adapter under a provider id. After registration,
 * routes may declare `provider: "<id>"` and `parseConfig` accepts the id
 * (the known-provider union includes it). Call at module init time:
 *
 *   // @ai-router/provider-bedrock
 *   registerAdapter("bedrock", () => new BedrockAdapter());
 */
export function registerAdapter(
  id: string,
  create: () => ProviderAdapter,
): void {
  if (id in builtinAdapters) {
    throw new Error(`registerAdapter: "${id}" collides with a built-in provider`);
  }
  if (!/^[a-z][a-z0-9_-]*$/i.test(id)) {
    throw new Error(`registerAdapter: "${id}" must match [a-zA-Z][a-zA-Z0-9_-]*`);
  }
  customAdapters.set(id, create());
}

/**
 * Every provider id valid in route configs: built-ins + preset ids +
 * registered custom ids. parseConfig validates against this union.
 */
export function knownProviderIds(): string[] {
  return [
    ...Object.keys(builtinAdapters),
    "openai-compatible",
    ...Object.keys(PROVIDER_PRESETS),
    ...customAdapters.keys(),
  ];
}

export function getAdapter(providerId: string): ProviderAdapter {
  const builtin =
    builtinAdapters[providerId] ??
    // Same wire protocol; only the base URL differs.
    (providerId === "openai-compatible" ? openai : undefined);
  if (builtin) return builtin;
  const custom = customAdapters.get(providerId);
  if (custom) return custom;
  throw new UnsupportedProviderError(
    `provider "${providerId}" is not implemented` +
      (PROVIDER_PRESETS[providerId]
        ? ""
        : ` (known: ${knownProviderIds().join(", ")})`),
  );
}

export function isSupported(providerId: string): boolean {
  return (
    providerId in builtinAdapters ||
    providerId === "openai-compatible" ||
    customAdapters.has(providerId)
  );
}

File: packages/python/src/ai_router/providers/registry.py
"""Adapter registry for built-in and third-party provider adapters."""

from __future__ import annotations

import re
from collections.abc import Callable
from typing import Any

from ..errors import UnsupportedProviderError
from .presets import PROVIDER_PRESETS

BUILTIN_PROVIDER_IDS: list[str] = [
    "openai",
    "azure",
    "anthropic",
    "gemini",
    "bedrock",
    "vertex",
]

_builtin_adapters: dict[str, Any] = {}
_custom_adapters: dict[str, Any] = {}


def register_builtin_adapter(id: str, adapter: Any) -> None:
    _builtin_adapters[id] = adapter


def register_adapter(id: str, create: Callable[[], Any]) -> None:
    """Register a third-party adapter under a provider id."""
    if id in BUILTIN_PROVIDER_IDS or id == "openai-compatible":
        raise ValueError(f'register_adapter: "{id}" collides with a built-in provider')
    if not re.match(r"^[a-zA-Z][a-zA-Z0-9_-]*$", id):
        raise ValueError(f'register_adapter: "{id}" must match [a-zA-Z][a-zA-Z0-9_-]*')
    _custom_adapters[id] = create()


def known_provider_ids() -> list[str]:
    """Every provider id valid in route configs: built-ins + preset ids + registered custom ids."""
    res: list[str] = list(BUILTIN_PROVIDER_IDS)
    res.append("openai-compatible")
    res.extend(list(PROVIDER_PRESETS.keys()))
    res.extend(list(_custom_adapters.keys()))
    return res


def get_adapter(provider_id: str) -> Any:
    # Ensure built-ins are loaded
    if not _builtin_adapters:
        from .anthropic import AnthropicAdapter
        from .azure import AzureAdapter
        from .bedrock import BedrockAdapter
        from .gemini import GeminiAdapter
        from .openai import OpenAIAdapter
        from .vertex import VertexAdapter

        _builtin_adapters["openai"] = OpenAIAdapter()
        _builtin_adapters["azure"] = AzureAdapter()
        _builtin_adapters["anthropic"] = AnthropicAdapter()
        _builtin_adapters["gemini"] = GeminiAdapter()
        _builtin_adapters["bedrock"] = BedrockAdapter()
        _builtin_adapters["vertex"] = VertexAdapter()

    lookup_id = "openai" if provider_id == "openai-compatible" else provider_id
    if lookup_id in _builtin_adapters:
        return _builtin_adapters[lookup_id]
    if provider_id in _custom_adapters:
        return _custom_adapters[provider_id]
    raise UnsupportedProviderError(
        f'provider "{provider_id}" is not implemented'
        + ("" if provider_id in PROVIDER_PRESETS else f" (known: {', '.join(known_provider_ids())})")
    )


def is_supported(provider_id: str) -> bool:
    lookup_id = "openai" if provider_id == "openai-compatible" else provider_id
    return lookup_id in BUILTIN_PROVIDER_IDS or provider_id in PROVIDER_PRESETS or provider_id in _custom_adapters
